### DISCUSSION TITLE
XAML Behavior vs. EventTriggerBehavior Action

### DISCUSSION THREAD

**Discussion by limeniye** [CommunityToolkit/Windows#290] [Q&A] [answered]
<h2>States</h2>

 - Create a simple project with the **CommunityToolkit.WinUI.UI.Behaviors**
 - Add the following code:
 
```C#
public class ShowAttachedFlyoutBehavior
	: BehaviorBase<FrameworkElement>
{
	protected override bool Initialize()
	{
		var result = ShowAttachedFlyout();
		return result;
	}

	protected override bool Uninitialize()
	{
		HideAttachedFlyout();
		return true;
	}

	private bool ShowAttachedFlyout()
	{
		if (AssociatedObject == null)
		{
			return false;
		}
		FlyoutBase.ShowAttachedFlyout(AssociatedObject);
		return true;
	}

	private void HideAttachedFlyout()
	{
		FlyoutBase.GetAttachedFlyout(AssociatedObject).Hide();
	}
}
```
```XML
<Grid x:Name="AddNewGrid" ColumnDefinitions="70,Auto,20" Height="{Binding ElementName=NavigationPanel, Path=ActualHeight}">
	<FlyoutBase.AttachedFlyout>
		<MenuFlyout>
			<MenuFlyoutItem Text="UDP"/>
			<MenuFlyoutItem Text="TCP"/>
		</MenuFlyout>
	</FlyoutBase.AttachedFlyout>
	<i:Interaction.Behaviors>
		<core:EventTriggerBehavior EventName="Tapped" SourceObject="{Binding ElementName=AddNewGrid}">
			<behaviors:ShowAttachedFlyoutBehavior/>
		</core:EventTriggerBehavior>
	</i:Interaction.Behaviors>
	<TextBlock Grid.Column="1" Text="New" VerticalAlignment="Center"/>
</Grid>
```
 - Deploy your application

<h2>Expectations</h2>

All should be without errors and warnings

<h2>Result</h2>

Parsing error
![image](https://github.com/unoplatform/uno/assets/69314237/43247a7f-87b0-4a59-9c97-09386c4179e0)

```
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
The text associated with this error code could not be found.
```

<h2>Option info</h2>

If you will remove the line ```<behaviors:ShowAttachedFlyoutBehavior/>``` the error will disappear.

Versions of **CommunityToolkit.WinUI.UI.Behaviors**: 7.1.2
Original post: https://github.com/unoplatform/uno/discussions/13145#discussioncomment-6657927
Visual Studio: 17.8.0 Preview 1.0
Repository: 
https://github.com/limeniye/UnoCustomBehaviorParsingError

[UnoCustomBehaviorParsingError-main.zip](https://github.com/unoplatform/uno/files/12343147/UnoCustomBehaviorParsingError-main.zip)

**Comment by limeniye** [comment]
@michael-hawker it steel doesn't work. The same error message.
Visual Studio 17.9.0 Preview 1.0
.Net 8.0
Uno.WinUI 5.0.**
CommunityToolkit.WinUI.Behaviors 8.0.230907

Maybe I'm doing something wrong?

**Reply by michael-hawker** [reply, marked as answer]
@limeniye your behavior setup doesn't make sense, I believe that's why you're having an issue. You'd probably get more info if you turned on mixed debugging, maybe?

It looks like you want to trigger an **Action** in response to the `EventTriggerBehavior`, so when tapped you open a flyout. For that you don't write a behavior, you use `IAction` from the XAML Behaviors contracts. That's what allows EventTriggerBehavior to call things in response.

Adding another behavior within there makes no sense. So, what's happening is your behavior is getting initialized right away and trying to open the flyout on load, which I assume is causing some issue within the platform code due to the timing of trying to open a flyout before the components are loaded?

You can look at similar examples of how the Animations system works in the toolkit with the StartAnimationAction.

**Reply by limeniye** [reply]
@michael-hawker, thank you for the your answer. Indeed, that works!
I apologize for my incompetence.

P.S.
>You'd probably get more info if you turned on mixed debugging

I've tried to enable the **Common Language Runtime Exceptions** but it gives me the same result.

![image](https://github.com/CommunityToolkit/Windows/assets/69314237/073049a8-e831-453e-9eb9-71bcc1633a77)

**Reply by michael-hawker** [reply]
No need to apologize, we're all learning all the time. 😊

If you right-click your project and go to **Properties**, you can go to the **Debug** settings here:
![image](https://github.com/CommunityToolkit/Windows/assets/24302614/9f6e3b0f-d6e8-42d3-a733-33e4b6121676)

You can open the _Launch Profiles UI_ and select **Enable native code debugging**:
![image](https://github.com/CommunityToolkit/Windows/assets/24302614/138502c9-5746-46cf-9c5b-7e0c2a1d7204)

This can sometimes give you different errors (C++/Win32) or details from exceptions that come from within the framework before they get bubbled up.

**Reply by limeniye** [reply]
@michael-hawker I appreciate your help. Thank you))
